### ISSUE TITLE
Ecallidraw getting slow in Obsidian

### ISSUE THREAD

**Issue by jpralle** [excalidraw/excalidraw#6474]
Hi,

I use Ecalidraw inside Obsidian (Android app v 1.4.3 with Excalidraw plugin v 1.8.17) on my Chromebook (ChromeOS 111.0.5563.118, Acer Spin 514) with a pen. I use it for handwritten Notes or small drawings. But once I write about one loosely filled screen of text, Excalidraw starts to become very slow. The sampling rate for the pen position becomes so slow that some letters are not written at all or become just a straight line (so just two pen positions are recorded by Excalidraw). 

I don't know how to give instructions to reproduce this or what information I can add to make this easier to understand or analyze, so please let me know and I'll try to provide whatever I can. I don't know what exactly the issue is here, but I have this related to the amount of text on a note just with Excalidraw in this setup. I did not experience it with other handwriting software (apps or web), only then the system load of my Chromebook is very high, which is not the case with Excalidraw, at least not when the note is pretty empty. I would also be happy with options to just "flatten" the drawing (meaning to make it just an image on which I can continue to write). I do not change stuff from the top of the nor and if I want to, I can just draw white lines above the old and rewrite, so that would be totally fine. But I didn't find an option to do that.

Regards,
Jan

**Comment by dwelle** [comment]
Hi! Does this happen (or happen to the same extent) on excalidraw.com as well, or only Obsidian?

**Comment by niposch** [comment]
I also experienced this general lagginess in writing/general slow down on relatively small handwritten notes. After around 20 mins of taking notes it gets to the point I have to start a new drawing because it's gotten too slow. I recreated it in this file with some example data: https://excalidraw.com/#json=okEmUOtTeNajW27iyorM9,qy7H15tM4a1OrGFqRxPVqA
The line gets further and further behind where the pen is on the screen as the note grows.
Even when all the filler data is off-screen, the performance is impacted as if all the example data was on screen.
I get the feeling that excalidraw shouldn't have to draw stuff that is off screen and instead use some sort of culling.
This can also be observed in the browser version of excalidraw. The performance is slightly better in the browser version, however it also gets bogged down easily. The obsidian version seems to be slower by around a factor of <=2.
I sometimes experience a delay where excalidraw registers my first pen position and then doesn't register additional pen positions for a short while, resulting in this:
![image](https://user-images.githubusercontent.com/45246069/232501273-3536f8be-654f-4a68-876c-61b06e2813cf.png)
This issue is fairly rare, but it also seems to happen on the web version. Albeit, it feels that it is rarer than in obsidian. This happens only on rather big canvases, so it might be related to performance issues.

**Comment by niposch** [comment]
@jpralle so you're using the android obsidian app on a chrome book?
Afaik you can install linux apps on your chrome book (obsidian is available as an appimage) maybe that'll somewhat improve performance for you.
=> https://beebom.com/install-appimage-programs-on-chromebook/

**Comment by jpralle** [comment]
Yes, I know. I had that installed, but it comes with its own difficulties: The pen pressure level is not available, so the drawings and writings get very sterile. Also it resides in a different part of the file system, which makes making files available harder. And the on-screen keyboard is not available. Thus I would try to avoid that. I'd rather load my drawings into excalidraw.com, I guess, than go back to the Linux app.

**Comment by dwelle** [comment]
We will attempt improving this soon!

**Comment by niposch** [comment]
I added a simple cache to the getElementBounds method. This alleviates all performance issues when zoomed in on a small area of the canvas. Not familiar with the code base, but caching this method seems like a no-brainer.
=> see https://github.com/niposch/excalidraw/tree/niposch-performance-hacks
I tried this on the huge canvas I linked earlier, and the general performance is just like in a new document when I have my view zoomed in like this:
![image](https://user-images.githubusercontent.com/45246069/233051216-540cf27d-3ec0-43e8-b000-9112bb0a6d53.png)
When the view is zoomed out, panning and zooming is still choppy.

**Comment by jpralle** [comment]
Hi,
any update on this?

**Comment by dwelle** [comment]
@jpralle first part was recently merged https://github.com/excalidraw/excalidraw/pull/6759

@niposch we've added a cache for this as well, and it should already be in Obsidian. Did you notice similar perf improvement?

https://github.com/excalidraw/excalidraw/blob/9e0bfd178e77e7a524496aa4fa22d81dd3b4e526/src/element/bounds.ts#L40-L46

**Comment by zsviczian** [comment]
I've not yet published an Obsidian update with #6759 because there are 2 open issues that I want to first resolve. 
#6885 and #6884

**Comment by dwelle** [comment]
> I've not yet published an Obsidian update with #6759 because there are 2 open issues that I want to first resolve. #6885 and #6884

I meant the boundCache (that's 2 months old).

**Comment by niposch** [comment]
On the example canvas I posted up top, it definitely works pretty well when I draw something off to the side. Feels like a fresh canvas when the other stuff isn't on-screen. Panning when zoomed in on the same area is also smooth.
Panning when really zoomed out is still quite choppy. 

Tested all this in the web version and the obsidian version, and it's basically the same.

Great work on this 🎉

**Comment by jpralle** [comment]
Great to hear! :) Though I did a quick test with an old note that had the problems and Obsidian for Android 1.4.6 but had only a small Improvement. Is the improvent in that setup I mentioned?
The small improvent still will help significantly, I guess, I just have to move most of the existing drawing off screen, which is a bit fiddly in every day use, but then it's significantly faster. So much thanks anyway :)

**Comment by nippyseven** [comment]
I am experiencing the same issue, where the sampling rate for the pen simply becomes unusable for writing unless I slow my writing to <20% speed, which is not feasible. I am using a Samsung Galaxy Tab S6 Lite (2022) with the latest version of Android, Obsidian, and the Excalidraw plugin. 

The lag increases dramatically the more pen strokes are in the document, but I have found that even on files with massive PDF documents imported, there is no noticeable difference in lag. Zooming in so there are no elements on screen helps but doesn't solve the problem and is not a viable long-term solution

While this is not a high-end device by any means, I think implementing an optimisation for the pen tool on mobile devices would be a massive improvement, especially when Excalidraw is essentially the only plugin that offers free drawing capabilities within Obsidian.  
I assume the lag is tied to RAM in some capacity, as my PC (16 GB DDR4 3600 Hz) has no lag whatsoever, my laptop (8 GB 2400 Hz) has some lag but is mostly usable on less dense files and my tablet (4 GB) is practically unusable with 20+ elements

**Comment by zerolfer** [comment]
Experiencing the same performance issue on MS Surface Pro 7. The pen is unusable. :(

**Comment by zsviczian** [comment]
I confess I did not read the whole thread. It is a known issue that the freedraw tool is not efficient for handwriting and I understand the drive to fix it.

As for Obsidian performance, please test it with the base OBsidian theme as well. Some of the themes - such as the popular minimal theme - adds CSS styling that has high impact on the DOM rendering performance and results in much slower Excalidraw experience.

**Comment by enzoptortelli** [comment]
Hey, if you're also having this problem, increasing the interval for autosaving did the trick for me. Hope it helps you as well!

**Comment by coyita1234** [comment]
Muchas gracias zsviczian, cuando había escrito mucho, había lag en excalidraw, eso no me permitía trabajar, pero luego leí tu comentario y quite el tema que descargue, puse el "por defecto" y el lag reducido muchísimo, gracias de verdad, sin embargo hay un poco de lag cuando escribo rápido, a veces si y a veces no, pero a comparación de antes hay una gran diferencia
